# <div class= "alert alert-success"> Analyzing DataFrames

In [1]:
# all libraries are imported at top
import pandas as pd       


In [3]:
path = "../bicycle_dataset/data.csv"
data_frame = pd.read_csv(path)    #loading csv file (bicycle ride dataset) to a DataFrame

# print(df.to_string())     # To view the whole DataFrame

###  <div class = "alert alert-info"> Info About the Data

    info() method gives us more information about the data set.

In [4]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9495235 entries, 0 to 9495234
Data columns (total 23 columns):
 #   Column             Dtype  
---  ------             -----  
 0   trip_id            int64  
 1   year               int64  
 2   month              int64  
 3   week               int64  
 4   day                int64  
 5   hour               int64  
 6   usertype           object 
 7   gender             object 
 8   starttime          object 
 9   stoptime           object 
 10  tripduration       float64
 11  temperature        float64
 12  events             object 
 13  from_station_id    int64  
 14  from_station_name  object 
 15  latitude_start     float64
 16  longitude_start    float64
 17  dpcapacity_start   float64
 18  to_station_id      int64  
 19  to_station_name    object 
 20  latitude_end       float64
 21  longitude_end      float64
 22  dpcapacity_end     float64
dtypes: float64(8), int64(8), object(7)
memory usage: 1.6+ GB


###  <div class = "alert alert-info"> Viewing the Data

####  <div class = "alert alert-danger">i) head( ) method : </div>
Common method for a quick overview of the DataFrame, is the head() method.The head() method returns the headers and a specified number of rows, starting from the top. 

In [5]:
print(data_frame.head(2))  # head shows specified number of rows of DataFrame from top. By default shows 5 rows

   trip_id  year  month  week  day  hour    usertype gender  \
0  2355134  2014      6    27    0    23  Subscriber   Male   
1  2355133  2014      6    27    0    23  Subscriber   Male   

             starttime             stoptime  ...  from_station_id  \
0  2014-06-30 23:57:00  2014-07-01 00:07:00  ...              131   
1  2014-06-30 23:56:00  2014-07-01 00:00:00  ...              282   

           from_station_name latitude_start  longitude_start dpcapacity_start  \
0  Lincoln Ave & Belmont Ave      41.939365       -87.668385             15.0   
1    Halsted St & Maxwell St      41.864580       -87.646930             15.0   

   to_station_id          to_station_name  latitude_end  longitude_end  \
0            303  Broadway & Cornelia Ave     41.945512     -87.645980   
1             22       May St & Taylor St     41.869482     -87.655486   

  dpcapacity_end  
0           15.0  
1           15.0  

[2 rows x 23 columns]


    If the number of rows is not specified, the head() method will return the top 5 rows.

In [6]:
print(data_frame.head())

   trip_id  year  month  week  day  hour    usertype  gender  \
0  2355134  2014      6    27    0    23  Subscriber    Male   
1  2355133  2014      6    27    0    23  Subscriber    Male   
2  2355130  2014      6    27    0    23  Subscriber    Male   
3  2355129  2014      6    27    0    23  Subscriber  Female   
4  2355128  2014      6    27    0    23  Subscriber  Female   

             starttime             stoptime  ...  from_station_id  \
0  2014-06-30 23:57:00  2014-07-01 00:07:00  ...              131   
1  2014-06-30 23:56:00  2014-07-01 00:00:00  ...              282   
2  2014-06-30 23:33:00  2014-06-30 23:35:00  ...              327   
3  2014-06-30 23:26:00  2014-07-01 00:24:00  ...              134   
4  2014-06-30 23:16:00  2014-06-30 23:26:00  ...              320   

             from_station_name latitude_start  longitude_start  \
0    Lincoln Ave & Belmont Ave      41.939365       -87.668385   
1      Halsted St & Maxwell St      41.864580       -87.646930   
2 

####  <div class = "alert alert-danger">ii) tail( ) method : 
The tail() method returns the headers and a specified number of rows, starting from the bottom. 

In [7]:
print(data_frame.tail(2))   # tail shows specified number of rows of DataFrame from bottom. By default shows 5 rows

          trip_id  year  month  week  day  hour    usertype  gender  \
9495233  16734067  2017     10    39    6     0  Subscriber  Female   
9495234  16734066  2017     10    39    6     0  Subscriber  Female   

                   starttime             stoptime  ...  from_station_id  \
9495233  2017-10-01 00:00:00  2017-10-01 00:06:00  ...              288   
9495234  2017-10-01 00:00:00  2017-10-01 00:12:00  ...              135   

                  from_station_name latitude_start  longitude_start  \
9495233  Larrabee St & Armitage Ave      41.918084       -87.643749   
9495234        Halsted St & 21st St      41.853780       -87.646650   

        dpcapacity_start  to_station_id            to_station_name  \
9495233             11.0            289      Wells St & Concord Ln   
9495234             11.0            129  Blue Island Ave & 18th St   

         latitude_end  longitude_end dpcapacity_end  
9495233     41.912133     -87.634656           19.0  
9495234     41.857556     -

    If the number of rows is not specified, the tail() method will return the bottom 5 rows.

In [8]:
print(data_frame.tail())

          trip_id  year  month  week  day  hour    usertype  gender  \
9495230  16734072  2017     10    39    6     0  Subscriber  Female   
9495231  16734071  2017     10    39    6     0  Subscriber    Male   
9495232  16734070  2017     10    39    6     0  Subscriber    Male   
9495233  16734067  2017     10    39    6     0  Subscriber  Female   
9495234  16734066  2017     10    39    6     0  Subscriber  Female   

                   starttime             stoptime  ...  from_station_id  \
9495230  2017-10-01 00:01:00  2017-10-01 00:12:00  ...              220   
9495231  2017-10-01 00:01:00  2017-10-01 00:12:00  ...              220   
9495232  2017-10-01 00:01:00  2017-10-01 00:15:00  ...              289   
9495233  2017-10-01 00:00:00  2017-10-01 00:06:00  ...              288   
9495234  2017-10-01 00:00:00  2017-10-01 00:12:00  ...              135   

                  from_station_name latitude_start  longitude_start  \
9495230  Hampden Ct & Diversey Pkwy      41.932620 

In [9]:
data_frame.describe()

trip_id          year         month          week           day  \
count  9.495235e+06  9.495235e+06  9.495235e+06  9.495235e+06  9.495235e+06   
mean   9.861173e+06  2.015737e+03  7.161242e+00  2.938068e+01  2.687387e+00   
std    4.680018e+06  1.075629e+00  2.708379e+00  1.177941e+01  1.890345e+00   
min    1.109427e+06  2.014000e+03  1.000000e+00  1.000000e+00  0.000000e+00   
25%    5.941660e+06  2.015000e+03  5.000000e+00  2.100000e+01  1.000000e+00   
50%    1.005810e+07  2.016000e+03  7.000000e+00  3.000000e+01  3.000000e+00   
75%    1.383260e+07  2.017000e+03  9.000000e+00  3.800000e+01  4.000000e+00   
max    1.753670e+07  2.017000e+03  1.200000e+01  5.300000e+01  6.000000e+00   

               hour  tripduration   temperature  from_station_id  \
count  9.495235e+06  9.495235e+06  9.495235e+06     9.495235e+06   
mean   1.362471e+01  1.144686e+01  6.299990e+01     1.799215e+02   
std    4.858830e+00  7.206061e+00  1.720086e+01     1.217041e+02   
min    0.000000e+00  2.000000e+00 -1.500000e+01     2.000000e+00   
25%    9.000000e+00  6.033333e+00  5.200000e+01     7.700000e+01   
50%    1.500000e+01  9.633333e+00  6.690000e+01     1.640000e+02   
75%    1.700000e+01  1.520000e+01  7.590000e+01     2.680000e+02   
max    2.300000e+01  6.000000e+01  9.500000e+01     6.260000e+02   

       latitude_start  longitude_start  dpcapacity_start  to_station_id  \
count    9.495235e+06     9.495235e+06      9.495235e+06   9.495235e+06   
mean     4.190017e+01    -8.764473e+01      2.138426e+01   1.807048e+02   
std      3.467450e-02     2.175210e-02      7.644496e+00   1.221144e+02   
min      4.173665e+01    -8.780287e+01      0.000000e+00   2.000000e+00   
25%      4.188103e+01    -8.765479e+01      1.500000e+01   7.700000e+01   
50%      4.189228e+01    -8.764107e+01      1.900000e+01   1.640000e+02   
75%      4.192008e+01    -8.763058e+01      2.300000e+01   2.730000e+02   
max      4.206431e+01    -8.754939e+01      5.500000e+01   6.260000e+02   

       latitude_end  longitude_end  dpcapacity_end  
count  9.495235e+06   9.495235e+06    9.495235e+06  
mean   4.190069e+01  -8.764496e+01    2.128447e+01  
std    3.494946e-02   2.181374e-02    7.595932e+00  
min    4.173665e+01  -8.780287e+01    0.000000e+00  
25%    4.188132e+01  -8.765549e+01    1.500000e+01  
50%    4.189384e+01  -8.764109e+01    1.900000e+01  
75%    4.192169e+01  -8.763058e+01    2.300000e+01  
max    4.206431e+01  -8.754939e+01    5.500000e+01

In [10]:
data_frame['usertype'].shape   # returns number of rows in 'usertype' column

(9495235,)

In [11]:
data_frame.columns   # attribute that shows column index

Index(['trip_id', 'year', 'month', 'week', 'day', 'hour', 'usertype', 'gender',
       'starttime', 'stoptime', 'tripduration', 'temperature', 'events',
       'from_station_id', 'from_station_name', 'latitude_start',
       'longitude_start', 'dpcapacity_start', 'to_station_id',
       'to_station_name', 'latitude_end', 'longitude_end', 'dpcapacity_end'],
      dtype='object')

In [12]:
data_frame['usertype']  # shows all data of usertype column

0          Subscriber
1          Subscriber
2          Subscriber
3          Subscriber
4          Subscriber
              ...    
9495230    Subscriber
9495231    Subscriber
9495232    Subscriber
9495233    Subscriber
9495234    Subscriber
Name: usertype, Length: 9495235, dtype: object

### <div class = "alert alert-danger"> How to Get Unique Values from a DataFrame Column?<div>

Consider a scenario where you have a dataframe consisting of students with respect to the subjects they study. In such a condition, if you wish to identify the total number of subjects being studied overall, it is quite infeasible for you to count each row of the dataframe and identify the unique subject category. To reduce your manual work, below are the two methods by which you can easily get unique values in a column of pandas dataframe:

- Using unique() method     
- Using value_counts() method

<div class = "alert alert-warning">     
unique() method for a series (column of DataFrame):</div>
    
unique() method is used when we deal with a single column of a DataFrame (i.e a series) and returns all unique elements of a column. The method returns a DataFrame containing the unique elements of a column, along with their corresponding index labels.
Uniques are returned in order of appearance. 

In [13]:
data_frame['usertype'].unique() # shows the value that could be assigned to each usertype i.e Subscriber,Customer or Dependent

#out of the 9495234 values, usertypr could be assign Subscriber,Customer or Dependent values

array(['Subscriber', 'Customer', 'Dependent'], dtype=object)

<div class = "alert alert-warning">     
 value_counts() method:</div>
 
The value_counts() method returns a Series containing the counts of unique values. This means, for any column in a dataframe, this method returns the count of unique entries in that column. Simply put value_counts() method returns occurences of unique value in a DataFrame column (series)

In [14]:
data_frame['usertype'].value_counts()

Subscriber    9493780
Customer         1277
Dependent         178
Name: usertype, dtype: int64

- Insight: Subscribed user are taking the most rides

### <div class = "alert alert-danger"> shape() method<div>
    
The shape of a DataFrame is a tuple that tells the number of rows and columns of a given DataFrame.

The DataFrame.shape attribute in Pandas enables us to obtain the shape of a DataFrame. For example, if a DataFrame has a shape of (80, 10) , this implies that the DataFrame is made up of 80 rows and 10 columns of data.

In [15]:
data_frame.shape  # returns (rows,columns)

(9495235, 23)

In [16]:
data_frame['gender'].shape

(9495235,)

In [17]:
data_frame['gender'].unique()

array(['Male', 'Female'], dtype=object)

### <div class = "alert alert-danger"> Filtering out rows<div>

In [18]:
data_frame['gender'] == 'Female'  # Say we are interestes in female riders only. This is a predicate

0          False
1          False
2          False
3           True
4           True
           ...  
9495230     True
9495231    False
9495232    False
9495233     True
9495234     True
Name: gender, Length: 9495235, dtype: bool

- row index against which there is True are femaol bike riders
- A predicate is a function that always returns True or False by performing some condition operations in a filter method. Simply put a boolean expression is called a predicate 

<div class = "alert alert-warning">     
To filter out Male riders so that only Female riders data is available: 
</div>

In [19]:
data_frame[data_frame['gender'] == 'Female']

trip_id  year  month  week  day  hour    usertype  gender  \
3         2355129  2014      6    27    0    23  Subscriber  Female   
4         2355128  2014      6    27    0    23  Subscriber  Female   
11        2355110  2014      6    27    0    22  Subscriber  Female   
13        2355100  2014      6    27    0    22  Subscriber  Female   
14        2355098  2014      6    27    0    22  Subscriber  Female   
...           ...   ...    ...   ...  ...   ...         ...     ...   
9495216  16734088  2017     10    39    6     0  Subscriber  Female   
9495219  16734084  2017     10    39    6     0  Subscriber  Female   
9495230  16734072  2017     10    39    6     0  Subscriber  Female   
9495233  16734067  2017     10    39    6     0  Subscriber  Female   
9495234  16734066  2017     10    39    6     0  Subscriber  Female   

                   starttime             stoptime  ...  from_station_id  \
3        2014-06-30 23:26:00  2014-07-01 00:24:00  ...              134   
4        2014-06-30 23:16:00  2014-06-30 23:26:00  ...              320   
11       2014-06-30 22:17:00  2014-06-30 22:22:00  ...              267   
13       2014-06-30 22:04:00  2014-06-30 22:08:00  ...              210   
14       2014-06-30 22:04:00  2014-06-30 22:24:00  ...              332   
...                      ...                  ...  ...              ...   
9495216  2017-10-01 00:07:00  2017-10-01 00:28:00  ...              488   
9495219  2017-10-01 00:06:00  2017-10-01 00:12:00  ...               62   
9495230  2017-10-01 00:01:00  2017-10-01 00:12:00  ...              220   
9495233  2017-10-01 00:00:00  2017-10-01 00:06:00  ...              288   
9495234  2017-10-01 00:00:00  2017-10-01 00:12:00  ...              135   

                  from_station_name latitude_start  longitude_start  \
3          Peoria St & Jackson Blvd      41.877749       -87.649633   
4          Loomis St & Lexington St      41.872187       -87.661501   
11          Lake Park Ave & 47th St      41.809443       -87.591875   
13        Ashland Ave & Division St      41.903448       -87.669313   
14       Halsted St & Diversey Pkwy      41.933341       -87.648747   
...                             ...            ...              ...   
9495216        Pulaski Rd & Eddy St      41.945636       -87.727737   
9495219             McCormick Place      41.851375       -87.618835   
9495230  Hampden Ct & Diversey Pkwy      41.932620       -87.642385   
9495233  Larrabee St & Armitage Ave      41.918084       -87.643749   
9495234        Halsted St & 21st St      41.853780       -87.646650   

        dpcapacity_start  to_station_id             to_station_name  \
3                   19.0            194        State St & Wacker Dr   
4                   15.0            134    Peoria St & Jackson Blvd   
11                  15.0            322       Kimbark Ave & 53rd St   
13                  19.0            350   Ashland Ave & Chicago Ave   
14                  15.0            350   Ashland Ave & Chicago Ave   
...                  ...            ...                         ...   
9495216             11.0            488        Pulaski Rd & Eddy St   
9495219             27.0            273      Michigan Ave & 18th St   
9495230             23.0             87  Racine Ave & Fullerton Ave   
9495233             11.0            289       Wells St & Concord Ln   
9495234             11.0            129   Blue Island Ave & 18th St   

         latitude_end  longitude_end dpcapacity_end  
3           41.887155     -87.627750           11.0  
4           41.877749     -87.649633           19.0  
11          41.799568     -87.594747           15.0  
13          41.895966     -87.667747           15.0  
14          41.895966     -87.667747           15.0  
...               ...            ...            ...  
9495216     41.945636     -87.727737           11.0  
9495219     41.857813     -87.624550           23.0  
9495230     41.925563     -87.658404           19.0  
9495233   

- Above data is only concerning female riders  

<div class = "alert alert-warning"> Suppose now you are interested in the attributes: 'year', 'month','starttime', 'stoptime' for female riders only:

In [20]:
data_frame[data_frame['gender'] == 'Female'][['year', 'month','starttime', 'stoptime']]

year  month            starttime             stoptime
3        2014      6  2014-06-30 23:26:00  2014-07-01 00:24:00
4        2014      6  2014-06-30 23:16:00  2014-06-30 23:26:00
11       2014      6  2014-06-30 22:17:00  2014-06-30 22:22:00
13       2014      6  2014-06-30 22:04:00  2014-06-30 22:08:00
14       2014      6  2014-06-30 22:04:00  2014-06-30 22:24:00
...       ...    ...                  ...                  ...
9495216  2017     10  2017-10-01 00:07:00  2017-10-01 00:28:00
9495219  2017     10  2017-10-01 00:06:00  2017-10-01 00:12:00
9495230  2017     10  2017-10-01 00:01:00  2017-10-01 00:12:00
9495233  2017     10  2017-10-01 00:00:00  2017-10-01 00:06:00
9495234  2017     10  2017-10-01 00:00:00  2017-10-01 00:12:00

[2378675 rows x 4 columns]

###  <div class = "alert alert-info"> Adding column in a data frame

In [21]:
data_frame["date"] = data_frame["year"]  #adding a date colum to dataframe thats equal to year (just like in dictionaries)

data_frame.columns   #  date column at the end

Index(['trip_id', 'year', 'month', 'week', 'day', 'hour', 'usertype', 'gender',
       'starttime', 'stoptime', 'tripduration', 'temperature', 'events',
       'from_station_id', 'from_station_name', 'latitude_start',
       'longitude_start', 'dpcapacity_start', 'to_station_id',
       'to_station_name', 'latitude_end', 'longitude_end', 'dpcapacity_end',
       'date'],
      dtype='object')

In [22]:
data_frame.columns.shape  # now we have 24 columns

(24,)

In [23]:
#Simple example
data_frame["date"] = "Dates"
data_frame.columns 
data_frame["date"]

0          Dates
1          Dates
2          Dates
3          Dates
4          Dates
           ...  
9495230    Dates
9495231    Dates
9495232    Dates
9495233    Dates
9495234    Dates
Name: date, Length: 9495235, dtype: object

###  <div class = "alert alert-info"> Info about dataset example

In [24]:
path = "./data1.csv"
df = pd.read_csv(path) 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 169 entries, 0 to 168
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Duration  169 non-null    int64  
 1   Pulse     169 non-null    int64  
 2   Maxpulse  169 non-null    int64  
 3   Calories  164 non-null    float64
dtypes: float64(1), int64(3)
memory usage: 5.4 KB


### Explaning the result of print(df.info()):

###### - RangeIndex: 169 entries, 0 to 168 --------------------------> Total number of rows = 169 (0 - 168)
###### - Data columns (total 4 columns):---------------------------->  Total vumber of columns/ attributes = 4
##### - The table:
| |Column|    Non-Null Count|  Dtype|  
|--|------|    --------------|  -----|  
| 0|Duration|  169 non-null|    int64|  
| 1|Pulse   |169 non-null|    int64|  
| 2|Maxpulse|169 non-null|    int64|  
| 3|Calories|164 non-null|    float64|

##### and dtypes: float64(1), int64(3) -------------> Show us the name of each column, with the data type

#### - Null Values:
        The info() method also tells us how many Non-Null values there are present in each column, and in our data set it 
        seems like there are 164 of 169 Non-Null values in the "Calories" column.

        Which means that there are 5 rows with no value at all, in the "Calories" column, for whatever reason.

        Empty values, or Null values, can be bad when analyzing data, and you should consider removing rows with empty 
        values. This is a step towards what is called cleaning data.